File: week1-array-string/problem11-문자열나누기/heejae.java
// 문제: 문자열 나누기
// 접근 방식: firstCount와 anotherCount의 개수를 센다. 
// 또한 첫번째 문자의 경우 firstChar에 해당 문자를 지정해둔 후 그 이후 문자들과 비교하여 firstCount 또는 anotherCount를 증가시키다.
// 분리가 안될 수도 있는 문자열 범위를 위해 totalLength 변수를 만들어 나중에 남게 된다면 answer를 하나 증가시켜준다.
// 시간복잡도: O(n)

class Solution {
    public int solution(String s) {
        int answer = 0;
        
        char firstChar = ' ';
        int firstCount = 0;
        int anotherCount = 0;
        int totalLength = 0;
        
        for(int i=0; i < s.length(); i++) {
            char c = s.charAt(i);
            
            if(firstCount == 0) {
                firstChar = c;
                firstCount++;
                continue;
            }
            
            if(c == firstChar) firstCount++;
            else anotherCount++;
            
            if(firstCount == anotherCount) {
                totalLength += (firstCount + anotherCount);
                answer++;
                
                firstCount = 0;
                anotherCount = 0;
            }
        }
        
        if(totalLength < s.length()) answer++;
        
        return answer;
    }
}

File: week2-math/problem08-하노이의탑/heejae.cpp
// 문제: 하노이의 탑
// 접근 방식: n-1개를 중간으로 옮긴다. -> 제일 큰 원판을 목적지로 옮긴다. -> n-1개를 다시 목적지로 옮긴다. (재귀)
// 시간복잡도: O(2ⁿ)

#include <string>
#include <vector>

using namespace std;

vector<vector<int>> answer;
void hanoi(int n, int from, int to);

vector<vector<int>> solution(int n) {
    
    hanoi(n, 1, 3);
     
    return answer;
}

void hanoi(int n, int from, int to) {
    
    int other = 6 - from - to;
    
    if(n == 1) {
        // 종료
        answer.push_back({ from, to });
        return;
    }
    
    // n-1개를 중간으로 옮김
    hanoi(n-1, from , other);
    
    // 제일 큰 원판을 to로 옮김
    answer.push_back({ from, to });
    
    // n-1개를 목적지로
    hanoi(n-1, other, to); 
}
